### PR TITLE
Add minter component and refactoring project

### PR DESCRIPTION
**Pull request by cloudvenger**

### CODE REVIEW

File: src/components/absorber/carbon.cairo [hunk 1]
```diff
@@ -15,6 +15,14 @@ mod AbsorberComponent {
 
     // Internal imports
     use carbon_v3::components::absorber::interface::IAbsorber;
+    use carbon_v3::components::absorber::interface::ICarbonCredits;
+
+    // Constants
+
+    const YEAR_SECONDS: u64 = 31556925;
+    const MULT_ACCURATE_SHARE: u256 = 1_000_000;
```
Comment (tekkac) [inline src/components/absorber/carbon.cairo:23]: We should add an Issue on finding out how to deal with this properly

File: src/components/absorber/carbon.cairo [hunk 2]
```diff
@@ -43,6 +51,10 @@ mod AbsorberComponent {
         value: u256
     }
 
+    mod Errors {
+        const INVALID_ARRAY_LENGTH: felt252 = 'ERC1155: invalid array length';
```
Comment (tekkac) [inline src/components/absorber/carbon.cairo]: not ERC1155-related

File: src/components/absorber/carbon.cairo [hunk 3]
```diff
@@ -166,6 +171,64 @@ mod AbsorberComponent {
         }
     }
 
+    #[embeddable_as(CarbonCreditsImpl)]
+    impl CarbonCredits<
+        TContractState, +HasComponent<TContractState>, +Drop<TContractState>
+    > of ICarbonCredits<ComponentState<TContractState>> {
+        fn get_cc_vintages(self: @ComponentState<TContractState>) -> Span<u256> {
+            let times = self.Absorber_times.read();
```
Comment (tekkac) [inline src/components/absorber/carbon.cairo:179]: Are we sure we want to use those times?

File: src/components/absorber/carbon.cairo [hunk 4]
```diff
@@ -166,6 +171,64 @@ mod AbsorberComponent {
...
+
+        fn compute_cc_distribution(
+            self: @ComponentState<TContractState>, share: u256
+        ) -> Span<u256> {
+            let times = self.Absorber_times.read();
+            let absorptions = self.Absorber_absorptions.read();
+            let absorptions_u256 = self.__span_u64_into_u256(absorptions.array().span());
+
+            // [Check] list time and absorptions are equal size
+            assert(times.len() == absorptions.len(), Errors::INVALID_ARRAY_LENGTH);
```
Comment (tekkac) [inline src/components/absorber/carbon.cairo:201]: Isn't this checked in set_absorptions?

File: src/components/absorber/carbon.cairo [hunk 5]
```diff
@@ -166,6 +171,64 @@ mod AbsorberComponent {
...
+                if index == 0 {
+                    current_absorption = *absorptions_u256[index];
+                } else {
+                    current_absorption = *absorptions_u256[index] - *absorptions_u256[index - 1];
+                }
+
+                cc_distribution
+                    .append(
+                        (current_absorption * share / MULT_ACCURATE_SHARE)
+                            .try_into()
```
Comment (tekkac) [inline src/components/absorber/carbon.cairo]: Why try_into? Aren't all types u256?
The multiplication overflow would not be caught be the expect.

File: src/components/minter/interface.cairo [hunk 6]
```diff
@@ -4,13 +4,15 @@ use starknet::ContractAddress;
...
-    fn get_available_value(self: @TContractState) -> u256;
-    fn get_claimed_value(self: @TContractState, account: ContractAddress) -> u256;
+    fn get_available_money_amount(self: @TContractState) -> u256;
+    fn get_max_money_amount(self: @TContractState) -> u256;
+    fn get_min_money_amount_per_tx(self: @TContractState) -> u256;
+    fn is_public_sale_open(self: @TContractState) -> bool;
     fn is_sold_out(self: @TContractState) -> bool;
     fn is_canceled(self: @TContractState) -> bool;
     fn set_public_sale_open(ref self: TContractState, public_sale_open: bool);
+    fn set_min_money_amount_per_tx(ref self: TContractState, min_money_amount_per_tx: u256);
```
Comment (tekkac) [inline src/components/minter/interface.cairo:15]: "money" is a bit vague here

File: src/components/absorber.cairo [hunk 7]
```diff
@@ -1,4 +1,4 @@
-mod module;
```
Comment (tekkac) [inline src/components/absorber.cairo:1]: We used to put all these in lib.cairo for clarity.

`lib.cairo`:
```
mod absorber {
  mod carbon;
  mod interface;
}
```

File: src/components/minter/mint.cairo [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    #[derive(Drop, starknet::Event)]
+    struct Buy {
+        #[key]
+        address: ContractAddress,
+        cc_vintage: Span<u256>,
+        cc_distributed: Span<u256>,
+    }
+
+    mod Errors {
+        const INVALID_ARRAY_LENGTH: felt252 = 'ERC1155: invalid array length';
```
Comment (tekkac) [inline src/components/minter/mint.cairo]: Mint vs ERC1155?

File: src/components/minter/mint.cairo [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+            let contract_address = get_contract_address();
+            let balance = erc20.balanceOf(contract_address);
+
+            // [Interaction] Transfer tokens
+            let caller_address = get_caller_address();
+            let success = erc20.transfer(caller_address, balance);
+            assert(success, 'Transfer failed');
+        }
+
+        fn transfer(
```
Comment (tekkac) [inline src/components/minter/mint.cairo]: Generally not a good idea to execute contracts from outside.
I assume this method be protected by Access Control later, but right now it's not.
Also, `transfer` might not be a precise enough name and clashes with the name of ERC20 methods.

File: src/components/minter/mint.cairo [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+            // [Interaction] Buy
+            self._buy(value, force)
+        }
+
+        fn set_min_money_amount_per_tx(
+            ref self: ComponentState<TContractState>, min_money_amount_per_tx: u256
+        ) {
+            // [Effect] Store value
+            self.Mint_min_money_amount_per_tx.write(min_money_amount_per_tx);
```
Comment (tekkac) [inline src/components/minter/mint.cairo:190]: Checks not done here

File: src/components/minter/mint.cairo [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        }
+
+        fn get_max_money_amount(self: @ComponentState<TContractState>) -> u256 {
+            self.Mint_max_money_amount.read()
```
Comment (tekkac) [inline src/components/minter/mint.cairo:194]: No checks done here either

**Review by tekkac — APPROVED** [review]
Minor comments, naming conventions, good pratices for reference.

lgtm :ok_hand:

File: src/components/absorber/carbon.cairo (hunk 1 above)
Comment (cloudvenger) [reply src/components/absorber/carbon.cairo:23]: Created an exploration Issue on in (explore from DeFi Smart Contract)

File: src/components/absorber/carbon.cairo (hunk 3 above)
Comment (cloudvenger) [reply src/components/absorber/carbon.cairo:179]: Need to create a specific Issue about how to extract Year of the Absorption Curve

File: src/components/absorber/carbon.cairo (hunk 5 above)
Comment (cloudvenger) [reply src/components/absorber/carbon.cairo]: good point, I need to remove it

File: src/components/minter/mint.cairo (hunk 8 above)
Comment (cloudvenger) [reply src/components/minter/mint.cairo]: need to be modify

File: src/components/minter/mint.cairo (hunk 9 above)
Comment (cloudvenger) [reply src/components/minter/mint.cairo]: Agreed and need to add Ownership control when we will add Access Control.

File: src/components/minter/mint.cairo (hunk 10 above)
Comment (cloudvenger) [reply src/components/minter/mint.cairo:190]: I will add it

**Pull request merged**